src/transaction/log_prior_recv.hpp

```diff
@@ -66,10 +69,12 @@ namespace cublog
 
       message_container m_messages;                           // internal queue of messages
       std::mutex m_messages_mutex;                            // protect access on message queue
-      std::condition_variable m_messages_condvar;             // notify internal thread of new messages
+      std::condition_variable m_messages_push_cv;             // notify internal thread of new messages
+      std::condition_variable m_messages_consume_cv;          // notify a set of message has been consumed
 
       std::thread m_thread;                                   // internal thread
       bool m_shutdown = false;                                // true to stop internal thread
+      bool m_paused = false;                                  // true to pause consuming messages
```

Comment:
What was the race condition.

Reply:
In `wait_until_empty_and_pause()`, it did not consider the shutdown of the `prior_recver`. It could wait until the queue is empty, which is not met if it's shutdown.